File: app/src/main/java/com/example/learnwordstrainer/repository/WordRepository.java
package com.example.learnwordstrainer.repository;

import android.app.Application;
import android.content.Context;

import com.example.learnwordstrainer.dao.WordDao;
import com.example.learnwordstrainer.db.WordDatabase;
import com.example.learnwordstrainer.model.Word;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class WordRepository {
    private final WordDao wordDao;
    private final ExecutorService executorService;

    public WordRepository(Context context) {
        WordDatabase db = WordDatabase.getDatabase(context);
        wordDao = db.wordDao();
        executorService = Executors.newSingleThreadExecutor();
    }

    public void addWord(String englishWord, String translation) {
        executorService.execute(() -> {
            Word word = new Word(englishWord, translation);
            wordDao.addWord(word);
        });
    }

    public Word getRandomWord() {
        try {
            return executorService.submit(wordDao::getRandomWord).get();
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public List<String> getRandomTranslations(int excludeId) {
        try {
            return executorService.submit(() -> wordDao.getRandomTranslations(excludeId)).get();
        } catch (Exception e) {
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    public boolean wordExists(String englishWord) {
        try {
            return executorService.submit(() -> wordDao.wordExists(englishWord)).get();
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    public int getWordCount() {
        try {
            return executorService.submit(wordDao::getWordCount).get();
        } catch (Exception e) {
            e.printStackTrace();
            return 0;
        }
    }

    public int getLearnedWordsCount() {
        try {
            return executorService.submit(wordDao::getLearnedWordsCount).get();
        } catch (Exception e) {
            e.printStackTrace();
            return 0;
        }
    }

    public void updateScore(int id, int correctCount, int wrongCount) {
        executorService.execute(() ->
                wordDao.updateScore(id, correctCount, wrongCount)
        );
    }
}

File: app/src/main/java/com/example/learnwordstrainer/domain/repository/WordRepository.kt
package com.example.learnwordstrainer.domain.repository

import com.example.learnwordstrainer.domain.model.Word
import kotlinx.coroutines.flow.Flow

interface WordRepository {
    suspend fun addWord(word: Word)

    suspend fun getRandomWord(): Word?

    suspend fun getRandomTranslations(excludeId: Int): List<String>

    suspend fun wordExists(sourceWord: String, sourceLanguageCode: String): Boolean

    suspend fun getWordCount(): Int

    suspend fun getLearnedWordsCount(): Int

    suspend fun updateScore(id: Int, correctCount: Int, wrongCount: Int)

    fun getWords(sourceLanguageCode: String): Flow<List<Word>>

    suspend fun deleteWord(word: Word)

    suspend fun getRandomWordForLanguage(languageCode: String): Word?

    suspend fun getWordForRepetition(sourceLanguageCode: String): Word?

    suspend fun getAnswerOptionsForWord(wordToRepeat: Word, targetLanguageCode: String): List<String>
}